### ISSUE TITLE
DDC-3986: [Paginator] Invalid generated query with usage of output walkers

### ISSUE THREAD

**Issue by doctrinebot** [doctrine/doctrine2#4846]
Jira issue originally created by user egeloen:

Hey!

I have some troube when I use the paginator with the output walkers enabled on MySQL 5.7. Here is my query that I pass to the paginator:

``` java
SELECT shop, Distance(shop.location, PointFromText(:location)) AS HIDDEN distance FROM Shopper\Component\Shop\Shop shop WHERE shop.location IS NOT NULL ORDER BY distance ASC
```

It uses some geospatial feature but the point is not here. When using the output walkers, the generated query is the following:

``` java
SELECT DISTINCT id*4 FROM (SELECT s0_.name AS name_0, ST_AsText(s0_.location)  AS location_1, s0_.created_at AS created_at_2, s0_.updated_at AS updated_at_3, s0_.id AS id_4, ST_Distance(s0_.location, ST_PointFromText(?)) AS sclr_5 FROM shop s0_ WHERE s0_.location IS NOT NULL) dctrn_result ORDER BY sclr*5 ASC LIMIT 10 OFFSET 0
```

Then, Mysql complains the following:

``` java
SQLSTATE[HY000]: General error: 3065 Expression #1 of ORDER BY clause is not in SELECT list, references column 'dctrn*result.sclr*5' which is not in SELECT list; this is incompatible with DISTINCT"
```

Am I missing something or is it a bug since when I disable the output walkers, the query is executable.

**Comment by mikemix** [comment]
Still not working for MySQL 5.7 :(

**Comment by WubbleWobble** [comment]
Not tested and no guarantees of completeness, but a quick fix looks like so:

    // LimitSubqueryOutputWalker.php:
    private function preserveSqlOrdering(array $sqlIdentifier, $innerSql, $sql, $orderByClause)
    {
        // If the sql statement has an order by clause, we need to wrap it in a new select distinct
        // statement
        if (! $orderByClause instanceof OrderByClause) {
            return $sql;
        }

        // Rebuild the order by clause to work in the scope of the new select statement
        /* @var array $orderBy an array of rebuilt order by items */
        $orderBy = $this->rebuildOrderByClauseForOuterScope($orderByClause);
        $orderByFields = str_replace([' DESC', ' ASC'], ['', ''], $orderBy);

        // Build the select distinct statement
        $sql = sprintf(
            'SELECT DISTINCT %s, %s FROM (%s) dctrn_result ORDER BY %s',
            implode(', ', $sqlIdentifier),
            implode(', ', $orderByFields),
            $innerSql,
            implode(', ', $orderBy)
        );

        return $sql;
    }

The two lines added (the ones that mention $orderByFields) add the order-by fields mentioned in the subquery into the select on the outer query as per MySQL's complaint.

**Comment by skonsoft** [comment]
Hello,

Any news ?

Thank you

**Comment by keywinf** [comment]
Yes, any news ?

**Comment by Ocramius** [comment]
See #5622 #5973

**Comment by binarious** [comment]
Is this fixed via https://github.com/doctrine/doctrine2/pull/6143?

**Comment by lcobucci** [comment]
> Is this fixed via #6143?

@binarious it should be, but we don't have a stable release for `v2.6.0` so you'll have to use `v2.6.x-dev` to use this feature for now.

**Comment by lcobucci** [comment]
I'll close this issue since it was handled via #6143, please reopen if you feel it's needed.
